src/Layer.js

```diff
@@ -154,71 +181,113 @@ export default class Layer {
+     *
+     * @param {carto.source} source - The new Source object.
+     * @param {carto.Viz?} viz - Optional. The new Viz object. Defaults to the current viz.
      * @memberof carto.Layer
      * @instance
      * @async
      * @api
      */
-    async update (source, viz) {
-        if (viz === undefined) {
-            // Use current viz
-            viz = this._viz;
-        }
+    async update (source, viz = this._viz) {
+        return this._update(source, viz, true);
+    }
+    async _update (source, viz, majorChange) {
         this._checkSource(source);
         this._checkViz(viz);
 
-        source = source._clone();
-        this._atomicChangeUID++;
-        const uid = this._atomicChangeUID;
+        if (source !== this._source) {
+            source = source._clone();
+        }
+
+        let uid;
+        if (majorChange) {
+            uid = { major: this._majorNextUID, minor: 0 };
+            this._majorNextUID++;
+        } else {
+            uid = { major: this._majorCurrentUID, minor: this._minorNextUID };
+            this._minorNextUID++;
+        }
+
         const loadImagesPromise = viz.loadImages();
-        const metadata = await source.requestMetadata(viz);
+
+        // Don't await directly to start requesting the images as soon as possible
+        const metadataPromise = source.requestMetadata(viz);
         await loadImagesPromise;
+        const metadata = await metadataPromise;
 
         await this._context;
-        if (this._atomicChangeUID > uid) {
-            throw new CartoRuntimeError('Another atomic change was done before this one committed (1)');
+
+        if (majorChange) {
+            if (this._majorCurrentUID > uid.major) {
+                throw new CartoRuntimeError(`Another \`Layer.update()\` finished before this one. Commit ${uid} overridden by commit ${this._majorCurrentUID}.`);
+            }
+        } else {
+            if (this._majorCurrentUID > uid.major) {
+                throw new CartoRuntimeError(`Another \`Layer.update()\` finished before this viz change. Commit ${uid} overridden by commit ${this._majorCurrentUID}.${this._minorCurrentUID}`);
+            }
+            if (this._minorCurrentUID > uid.minor) {
+                throw new CartoRuntimeError(`Another \`viz change\` finished before this one. Commit ${uid.major}.${uid.minor} overridden by commit ${this._majorCurrentUID}.${this._minorCurrentUID}`);
+            }
         }
+        this._majorCurrentUID = uid.major;
+        this._minorCurrentUID = uid.minor;
 
         // Everything was ok => commit changes
         this.metadata = metadata;
 
-        source.bindLayer(this._onDataframeAdded.bind(this), this._onDataLoaded.bind(this));
-
-        if (this._source !== source) {
-            this._freeSource();
-        }
-
-        this._source = source;
-        this._noFirstRequestData = false;
-        this.requestData();
-
         viz.setDefaultsIfRequired(this.metadata.geomType);
```

Comment:
This block of sentences regarding to updating the viz can be easily extracted to a new `_updateViz()` to reduce a bit the size of this method

Reply:
I think that would suggest that the new method can be used without the `_update` method when the source doesn't change, but that is no true since a change in the `viz` may require new re-instantiations in the source.

Reply:
Even with just one proper use from _update, I would extract it to reduce the number of lines, but it's up to you